File: src/main/java/net/atos/zac/app/util/RESTExceptionMapper.java
/*
 * SPDX-FileCopyrightText: 2021 - 2022 Atos
 * SPDX-License-Identifier: EUPL-1.2+
 */

package net.atos.zac.app.util;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.logging.Level;
import java.util.logging.Logger;

import jakarta.ws.rs.WebApplicationException;
import jakarta.ws.rs.core.MediaType;
import jakarta.ws.rs.core.Response;
import jakarta.ws.rs.ext.ExceptionMapper;
import jakarta.ws.rs.ext.Provider;

import org.apache.commons.lang3.exception.ExceptionUtils;

import com.fasterxml.jackson.databind.ObjectMapper;

/**
 * Implementatie van ExceptionMapper. Alle exceptions worden gecatched door de JAX-RS runtime en gemapped naar een {@link Response}.
 */
@Provider
public class RESTExceptionMapper implements ExceptionMapper<Exception> {

    private static final Logger LOG = Logger.getLogger(RESTExceptionMapper.class.getName());

    /**
     * Retourneert een {@link Response} naar de Angular client.
     */
    @Override
    public Response toResponse(final Exception e) {
        if (e instanceof WebApplicationException &&
            Response.Status.Family.familyOf(((WebApplicationException) e).getResponse().getStatus()) !=
                                                    Response.Status.Family.SERVER_ERROR) {
            final WebApplicationException wae = (WebApplicationException) e;
            return Response.status(wae.getResponse().getStatus())
                    .type(MediaType.APPLICATION_JSON)
                    .entity(getJSONMessage(e, wae.getMessage()))
                    .build();
        } else {
            LOG.log(Level.SEVERE, e.getMessage(), e);
            return Response.status(Response.Status.INTERNAL_SERVER_ERROR)
                    .type(MediaType.APPLICATION_JSON)
                    .entity(getJSONMessage(e, "Algemene Fout"))
                    .build();
        }
    }

    private static String getJSONMessage(final Exception e, final String melding) {
        final Map<String, Object> data = new HashMap<>();
        data.put("message", melding);
        data.put("exception", e.getMessage());
        data.put("stackTrace", ExceptionUtils.getStackTrace(e));
        final ObjectMapper mapper = new ObjectMapper();
        try {
            return mapper.writeValueAsString(data);
        } catch (final IOException ioe) {
            //Het omzetten van de exceptie naar een JSON bericht is fout gegaan.
            LOG.severe(ioe.getMessage());
        }
        return null;
    }
}

File: src/main/java/net/atos/zac/app/util/RESTExceptionMapper.kt
/*
 * SPDX-FileCopyrightText: 2021 - 2022 Atos
 * SPDX-License-Identifier: EUPL-1.2+
 */
package net.atos.zac.app.util

import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.ws.rs.WebApplicationException
import jakarta.ws.rs.core.MediaType
import jakarta.ws.rs.core.Response
import jakarta.ws.rs.ext.ExceptionMapper
import jakarta.ws.rs.ext.Provider
import org.apache.commons.lang3.exception.ExceptionUtils
import java.io.IOException
import java.util.logging.Level
import java.util.logging.Logger

/**
 * Implementatie van ExceptionMapper. Alle exceptions worden gecatched door de JAX-RS runtime en gemapped naar een [Response].
 */
@Provider
class RESTExceptionMapper : ExceptionMapper<Exception> {
    companion object {
        private val LOG = Logger.getLogger(RESTExceptionMapper::class.java.name)
        const val JSON_CONVERSION_ERROR_MESSAGE = "Failed to convert exception to JSON"
    }

    /**
     * Retourneert een [Response] naar de client.
     */
    override fun toResponse(exception: Exception): Response {
        if (exception is WebApplicationException &&
            Response.Status.Family.familyOf(exception.response.status) != Response.Status.Family.SERVER_ERROR
        ) {
            return Response.status(exception.response.status)
                .type(MediaType.APPLICATION_JSON)
                .entity(getJSONMessage(exception, exception.message))
                .build()
        } else {
            LOG.log(Level.SEVERE, exception.message, exception)
            return Response.status(Response.Status.INTERNAL_SERVER_ERROR)
                .type(MediaType.APPLICATION_JSON)
                .entity(getJSONMessage(exception, "Algemene Fout"))
                .build()
        }
    }

    private fun getJSONMessage(exception: Exception, melding: String?): String {
        val data: MutableMap<String, Any?> = HashMap()
        melding?.let { data["message"] = it }
        data["exception"] = exception.message
        data["stackTrace"] = ExceptionUtils.getStackTrace(exception)
        return try {
            ObjectMapper().writeValueAsString(data)
        } catch (ioException: IOException) {
            LOG.log(Level.SEVERE, JSON_CONVERSION_ERROR_MESSAGE, ioException)
            JSON_CONVERSION_ERROR_MESSAGE
        }
    }
}
